File: code/js/src/http/NavigationRepository.ts
import { SirenAction, SirenLink } from "./Siren"


export class NavigationRepository implements NavigationRepository {

    linkStorage = new Map<string, SirenLink | null>()
    actionStorage = new Map<string, SirenAction | null>()


    getLink(key: string): SirenLink | null {
        return this.linkStorage.get(key)
    }

    getAction(key: string): SirenAction | null {
        return this.actionStorage.get(key)
    }

    addLinks(keys: string[], links: SirenLink[] | null) {
        if (links == null) return
        keys.forEach(key => {
            this.linkStorage.set(key, links?.find(link => link.rel.includes(key)))
        })
    }

    addActions(keys: string[], actions: SirenAction[] | null) {
        if (actions == null) return
        keys.forEach(key => {
            this.actionStorage.set(key, actions?.find(action => action.title == key))
        })
    }

    async ensureLink(key: string, fetchLink: () => any): Promise<SirenLink> {
        if (this.getLink(key) == null) {
            await fetchLink()
        }
        return this.getLink(key)
    }

    async ensureAction(key: string, fetchAction:() => any): Promise<SirenAction> {
        if (this.getAction(key) == null) {
            await fetchAction()
        }
        return this.getAction(key)
    }
}

File: code/android/app/src/main/java/isel/ps/classcode/http/NavigationRepository.kt
package isel.ps.classcode.http

import isel.ps.classcode.http.hypermedia.SirenAction
import isel.ps.classcode.http.hypermedia.SirenLink

class NavigationRepository {
    private val linkStorage = mutableMapOf<String, SirenLink?>()
    private val actionsStorage = mutableMapOf<String, SirenAction?>()

    private fun addLink(key: String, link: SirenLink?) = linkStorage.put(key, link)
    private fun addAction(key: String, action: SirenAction?) = actionsStorage.put(key, action)
    private fun getLink(key: String): SirenLink? = linkStorage[key]
    private fun getAction(key: String): SirenAction? = actionsStorage[key]

    fun addLinks(keys: List<String>, links: List<SirenLink>?) {
        if (links == null) return
        keys.forEach { key -> addLink(key = key, link = links.find { it.rel.contains(key) }) }
    }

    fun addActions(keys: List<String>, actions: List<SirenAction>?) {
        if (actions == null) return
        keys.forEach { key -> addAction(key = key, action = actions.find { it.title == key }) }
    }

    suspend fun ensureLink(key: String, fetchLink: suspend () -> Any): SirenLink? {
        val link = getLink(key)
        if (link == null) {
            fetchLink()
        }
        return link
    }

    suspend fun ensureAction(key: String, fetchAction: suspend () -> Any): SirenAction? {
        if (getAction(key) == null) {
            fetchAction()
        }
        return getAction(key)
    }
}
